Pynest microcircuit

hendrikrth opened the pull request:
The example files of NEST are lacking a PyNEST version of the microcircuit #450. To compensate this I would like you to include my branch. As reviewers I suggest @jhnnsnk and @mschmidt87.

jhnnsnk commented on pynest/examples/pynest-microcircuit/readme.md:
```diff
@@ -0,0 +1,29 @@
+May 2016
+
+## Description ##
+This is a PyNEST implementation of the SLI cortical microcircuit from the microcircuit model by Potjans and Diesmann (2014): The cell-type specific
```
This is a PyNEST implementation of the microcircuit model ...
(no need to mention SLI)

jhnnsnk commented on pynest/examples/pynest-microcircuit/readme.md:
```diff
@@ -0,0 +1,29 @@
+
+* Files:
+	* network.py  
+	file which gathers all parameters and connects the different nodes with each other
```
remove 'file which' to be consistent

mschmidt87 commented on pynest/examples/pynest-microcircuit/readme.md:
```diff
@@ -0,0 +1,29 @@
+	contains the parameters for the stimuli
+	* example.py  
+   use this script to try out the microcircuit
+   
```
- Please put a '.' behind each sentence.
- Please start the lase sentence with a capital letter: "Use this script..."

jhnnsnk commented on pynest/examples/pynest-microcircuit/readme.md:
```diff
@@ -0,0 +1,29 @@
+   
+How to use the example:
+
+To run the microcircuit on a local machine, adjust the variables N\_scaling and K\_scaling in network_params.py to 0.1 and create an output directory called '/data'. Then run 'example.py' with python. The output will be saved in the directory '/data'.
```
give an explanation what N_scaling and K_scaling mean and state that the full network refers to the value 1

mschmidt87 commented on pynest/examples/pynest-microcircuit/readme.md:
```diff
@@ -0,0 +1,29 @@
+
+To run the microcircuit on a local machine, adjust the variables N\_scaling and K\_scaling in network_params.py to 0.1 and create an output directory called '/data'. Then run 'example.py' with python. The output will be saved in the directory '/data'.
+The code can be parallelized using OpenMP and MPI, if NEST has been build with these applications. The number of threads (per MPI process) can be chosen by adjusting local\_num\_threads in sim_params.py. The number of MPI process can be set by running the script with the command 'mpirun -n num\_MPI\_processes python example.py'.
+
```
Please put the variable names in code blocks to improve the formatting, like so: `N_scaling`.

mschmidt87 commented on pynest/examples/pynest-microcircuit/readme.md:
```diff
@@ -0,0 +1,29 @@
+How to use the example:
+
+To run the microcircuit on a local machine, adjust the variables N\_scaling and K\_scaling in network_params.py to 0.1 and create an output directory called '/data'. Then run 'example.py' with python. The output will be saved in the directory '/data'.
+The code can be parallelized using OpenMP and MPI, if NEST has been build with these applications. The number of threads (per MPI process) can be chosen by adjusting local\_num\_threads in sim_params.py. The number of MPI process can be set by running the script with the command 'mpirun -n num\_MPI\_processes python example.py'.
```
- should read "built".
- A link to the documentation page would be useful here: [http://www.nest-simulator.org/parallel_computing/](http://www.nest-simulator.org/parallel_computing/)

mschmidt87 commented on pynest/examples/pynest-microcircuit/readme.md:
```diff
@@ -0,0 +1,29 @@
+The code can be parallelized using OpenMP and MPI, if NEST has been build with these applications. The number of threads (per MPI process) can be chosen by adjusting local\_num\_threads in sim_params.py. The number of MPI process can be set by running the script with the command 'mpirun -n num\_MPI\_processes python example.py'.
+
+Tested configuration:
+This version has been tested with NEST 2.10.0, Python 2.7.12, NumPy 1.11.1
```
Please insert the missing '.' at the end.

jhnnsnk commented on pynest/examples/pynest-microcircuit/example.py:
```diff
@@ -0,0 +1,47 @@
+Hendrik Rothe, Hannah Bos, Sacha van Albada; May 2016
+'''
+
+from network_params import *
```
to avoid `import *`, use for example `from network_params import net_dict` (the other import accordingly)

jhnnsnk commented on pynest/examples/pynest-microcircuit/example.py:
```diff
@@ -0,0 +1,47 @@
+# connect all nodes
+t1 = time.time()
+net.connect()
+print 'time to create the connections', time.time()-t1
```
- specify seconds as unit
- format the number, e.g. 2 decimals
- 'connections:'
- for consistency, start all printouts with capital letters

mschmidt87 commented on pynest/examples/pynest-microcircuit/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+'''
+pynest microcicuit simulation parameters
```
--> microcircuit

mschmidt87 commented on pynest/examples/pynest-microcircuit/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+sim_dict = {
+    # Simulation time (in ms)
+    't_sim': 1000.0,
+    # sim_resolution = Resolution of the simulation (in ms)
```
Please make the explanations consistent, so remove the 'sim_resolution = ' here.

mschmidt87 commented on pynest/examples/pynest-microcircuit/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+
+import os
+current_dir = os.getcwd()
+full_path = current_dir + '/data/'
```
Please use `os.path.join` to concatenate the paths.

mschmidt87 commented on pynest/examples/pynest-microcircuit/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+    'master_seed': 55,
+    # Number of threads run by NEST
+    'local_num_threads': 1,
+    # Recording intervall of the membrane potential (in ms)
```
--> interval

mschmidt87 commented on pynest/examples/pynest-microcircuit/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+    'local_num_threads': 1,
+    # Recording intervall of the membrane potential (in ms)
+    'rec_V_int': 1.0,
+    # If True data will be overwritten, if False it won't
```
should be: "If True, data will be overwritten. If False, a NESTError is raised if the files already exist."

mschmidt87 commented on pynest/examples/pynest-microcircuit/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+    'data_path': full_path,
+    # Masterseed for NEST and NumPy
+    'master_seed': 55,
+    # Number of threads run by NEST
```
--> "Number of threads per MPI process"

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+def get_std_delays(std_delay_exc, std_delay_inh, number_of_pop):
+    """Creates matrix containing the standard deviations of all delays.
+
+    Arguments
```
Arguments:

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+     evoked postsynaptic potential.
+
+    Arguements:
+    PSP_rel: standard deviation of the evoked synaptiv potential
```
synaptic

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    inh = PSP_e * g
+    weights[:, 0:dim:2] = exc
+    weights[:, 1:dim:2] = inh
+    # The weight of the connection L23E to L4E is doubled.
```
This comment should be contained in the docsctrings of the function.

jhnnsnk commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+
+
+def get_mean_delays(mean_delay_exc, mean_delay_inh, number_of_pop):
+    """Creates matrix containing the delay of all connections.
```
docstring conventions, e.g., as used by topology (compare numpy)

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    std_value_normal = PSP_rel
+    std_value_exception = 0.05
+    std_mat[:, :] = std_value_normal
+    std_mat[0, 2] = std_value_exception
```
What is the reason for doing this? We find this neither in the SLI implementation nor in the paper.

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    """This function creates a matrix of the standard deviation of the
+     evoked postsynaptic potential.
+
+    Arguements:
```
Arguments:

jhnnsnk commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    # The default recording device is the spike_detector. If you also
+    # want to record the membrane potentials of the neurons, add
+    # 'voltmeter' to the list.
+    'rec_dev': ['spike_detector', 'voltmeter'],
```
either remove 'voltmeter' from the list or adapt the documentation

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    'N_scaling': 0.1,
+    # Amplitude of excitatory postsynaptic potential.
+    'PSP_e': 0.15,
+    # Standard deviation of the postsynaptic potential.
```
Should be "Relative standard deviation.."

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    'mean_delay_matrix': get_mean_delays(1.5, 0.75, 8),
+    # std delay matrix.
+    'std_delay_matrix': get_std_delays(0.75, 0.375, 8),
+    # Rate of the poissonian spike generator.
```
--> "Poissonian"

jhnnsnk replied:
Define all values explicitly in the net_dict (e.g. `g` is missing).
I suggest to update the dictionary afterwards with the calculated dependent parameters.
Avoid to define the same variable twice like PSP_e and the first argument of PSP_mean_matrix().

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    'bg_rate': 8.,
+    # Turn Poisson input on or off (True or False).
+    'poisson_input': True,
+    # Delay of the poisson generator
```
Poisson

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    'poisson_input': True,
+    # Delay of the poisson generator
+    'poisson_delay': 1.5,
+    # Parameter of the neurons.
```
Parameters

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+        'E_L': -65.0,
+        # Threshold potential of the neurons (in mV).
+        'V_th': -50.0,
+        # Membrane potential after a spike (in mV).
```
Reset membrane potential...

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+        'tau_m': 10.0,
+        # Time constant of postsynaptic excitatory currents (in ms).
+        'tau_ex': 0.5,
+        # Time constant of postsynaptic inhibitatory currents (in ms).
```
inhibitory

jhnnsnk commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    # Factor to scale the number of neurons.
+    'N_scaling': 0.1,
+    # Amplitude of excitatory postsynaptic potential.
+    'PSP_e': 0.15,
```
unit missing

jhnnsnk commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    # std delay matrix.
+    'std_delay_matrix': get_std_delays(0.75, 0.375, 8),
+    # Rate of the poissonian spike generator.
+    'bg_rate': 8.,
```
unit missing

jhnnsnk commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    # Turn Poisson input on or off (True or False).
+    'poisson_input': True,
+    # Delay of the poisson generator
+    'poisson_delay': 1.5,
```
unit missing

jhnnsnk commented on pynest/examples/pynest-microcircuit/stimulus_params.py:
```diff
@@ -0,0 +1,63 @@
+    # Duration of the thalamic input (in ms)
+    'th_duration': 10.0,
+    # Rate of the thalamic input
+    'th_rate': 80.0,
```
unit missing

jhnnsnk replied:
- The thalamic rate in the SLI microcircuit is 120 1/s. What is the reason for 80 here?
- The default values for the time interval of thalamic input could be the same for the SLI and the PyNEST version.

jhnnsnk commented on pynest/examples/pynest-microcircuit/stimulus_params.py:
```diff
@@ -0,0 +1,63 @@
+    # Delay of the thalamic input (in ms)
+    'delay_th': np.asarray([1.5 for i in range(8)]),
+    # Standard deviation of the thalamic delay
+    'delay_th_sd': np.asarray([0.75 for i in range(8)]),
```
unit missing

jhnnsnk commented on pynest/examples/pynest-microcircuit/helpers.py:
```diff
@@ -0,0 +1,142 @@
+    """This function computes the weight which elicits a change in the membrane
+    potential of size PSP_val.
+
+    Arguements:
```
Arguments:

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+    # Delay of the poisson generator
+    'poisson_delay': 1.5,
+    # Parameter of the neurons.
+    'n_params': {
```
Here 'neuron_params' would be a less ambiguous name, since the 'n' could also stand for 'network'.

jhnnsnk commented on pynest/examples/pynest-microcircuit/helpers.py:
```diff
@@ -0,0 +1,142 @@
+
+def get_weight(PSP_val):
+    """This function computes the weight which elicits a change in the membrane
+    potential of size PSP_val.
```
elaborate a bit that a transformation from a voltage (PSP) to a current (PSC) is meant

jhnnsnk commented on pynest/examples/pynest-microcircuit/helpers.py:
```diff
@@ -0,0 +1,142 @@
+    scaling = net_dict['N_scaling']
+    N_full = net_dict['N_full']
+    conn_probs = stim_dict['conn_probs_th']
+    scaling = net_dict['N_scaling']
```
lines 94 and 97 are the same

mschmidt87 commented on pynest/examples/pynest-microcircuit/network_params.py:
```diff
@@ -0,0 +1,192 @@
+        # Membrane time constant (in ms).
+        'tau_m': 10.0,
+        # Time constant of postsynaptic excitatory currents (in ms).
+        'tau_ex': 0.5,
```
Here, 'tau_syn_ex' and 'tau_syn_in' would be better to keep these parameter names consistent with the NEST names. If you keep these parameter names strictly consistent, then you can just pass this dictionary to `nest.SetStatus` in the `Network` class, where you set the neuron parameters.
For this, you have to define `V0_mean` and `V0_std` outside of this dictionary.

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                    [gid],
+                    {'V_m': self.pyrngs[vp].uniform(
+                                self.net_dict['n_params']['V_th'],
+                                self.net_dict['n_params']['V_reset'])})
```
Here, the initial membrane potentials are always distributed uniformly, instead of using `V0_mean` and `V0_std`, which are defined `networks_params.py`. Either remove these parameters then, if they are not used anyway, or use them here.
The SLI version uses a normal distribution with customizable mean and sigma, so it would be good to have both versions consistent.

mschmidt87 commented on pynest/examples/pynest-microcircuit/helpers.py:
```diff
@@ -0,0 +1,142 @@
+    """
+    N_full = net_dict['N_full']
+    conn_probs = net_dict['conn_probs']
+    scaling = net_dict['N_scaling']
```
Why don't you use the `K_scaling` parameter here?

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                self.K_scl,
+                self.weight_mat, self.w_from_PSP,
+                self.DC_amp_e)
+            print 'The number of synapses is scalled by a factor of:'
```
scaled

mschmidt87 commented on pynest/examples/pynest-microcircuit/helpers.py:
```diff
@@ -0,0 +1,142 @@
+    scaling factor for the connections (K_scaling),
+    weight matrix, created by the function 'weight_mat' (w),
+    DC input, vector with zeros (DC).
+
```
This list does not match with the arguments of the function.

mschmidt87 commented on pynest/examples/pynest-microcircuit/helpers.py:
```diff
@@ -0,0 +1,142 @@
+    w_mean = w_from_PSP
+    K_ext = net_dict['K_ext']
+    bg_rate = net_dict['bg_rate']
+    internal_scaling = K_scaling
```
It seems that the distinction between `internal_scaling` and `external_scaling` is superfluous here.
You can just use `K_scaling` without creating two new variables.

mschmidt87 commented on pynest/examples/pynest-microcircuit/helpers.py:
```diff
@@ -0,0 +1,142 @@
+def adj_w_ext_to_K(K_full, K_scaling, w, w_from_PSP, DC):
+    """With this funtion the recurrent and external weights are adjusted
+    to the scaling of the indegrees. Extra DC input is added to
+    compensate the scaling and preserve the mean and variance of the input.
```
- The documentation is not complete because the internal weights are adapted as well.
- The documentation will become clearer if you add "Extra DC input I_ext is added"

jhnnsnk commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+
+    # create neuronal populations
+    def create_populations(self):
+        self.N_ttl = self.net_dict['N_full']
```
For clarity, I would recommend to name the parameters as in the dictionary, e.g., `self.N_full = self.net_dict['N_full']`.

mschmidt87 commented on pynest/examples/pynest-microcircuit/helpers.py:
```diff
@@ -0,0 +1,142 @@
+def get_total_number_of_synapses():
+    """This function returns the total number of synapses between
+    all populations.
+
```
Please make clear in the docstring, that this function is an internal helper function that ignores the `K_scaling` parameter.

jhnnsnk commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+        self.weight_mat_std = self.net_dict['PSP_std_matrix']
+        self.w_ext = self.w_from_PSP
+        self.DC_amp_e = np.zeros(len(self.net_dict['populations']))
+        if self.N_scl != 1:
```
I suggest to print both scaling factors always and not only if they are != 1.

jhnnsnk commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                self.DC_amp_e)
+            print 'The number of synapses is scalled by a factor of:'
+            print self.K_scl
+        else:
```
else is redundant here

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+import numpy as np
+from helpers import *
+import matplotlib.pyplot as plt
+import os
```
`os` is never used.

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+        if stim_dict is not None:
+            self.stim_dict = stim_dict
+        print 'Data will be written to %s' % self.data
+
```
All members of the class should be initialized in the `__init__` function and not in the member functions.

mschmidt87 commented on pynest/examples/pynest-microcircuit/helpers.py:
```diff
@@ -0,0 +1,142 @@
+    rate of the poissonian spike generator (bg_rate),
+    scaling factor for the connections (K_scaling),
+    weight matrix, created by the function 'weight_mat' (w),
+    DC input, vector with zeros (DC).
```
This argument should handle the case that `net_dict['poisson_input']=False`, i.e., where the Poisson input is replaced by constant DC input.

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                            })
+            node_info = nest.GetStatus(population)
+            local_nodes = [(ni['global_id'], ni['vp'])
+                           for ni in node_info if ni['local']]
```
There seems to be no function `GetLocalNodes` in PyNEST. However, 
`nest.sli_run('0 << /model  /%s >> GetLocalNodes' % net_dic['neuron_model'])`
`local_nodes = nest.sli_pop()`

should do the job.

jhnnsnk commented:
Thank you for this great work! Don't worry - most of our comments are just minor issues. :-)
Next week, we will continue looking through the code.

General remark:
The PyNEST and SLI microcircuit versions should have a similar folder name (currently: pynest_microcircuit vs. Potjans_2014)
In the respective readme files, a reference to the other version should be added.

mschmidt87 commented:
Thanks a lot for this great work.

- In general, it would be good to have a common policy for the helper functions:
Either the function gets all necessary passed as arguments or it retrieves all parameters from `net_dict`. `net_dict` can then be passed to the function. 
Currently, this is handled differently, e.g. in `synapses_th_matrix` and `adj_w_ext_to_K`.

- There is the option to switch off Poisson input. However, no DC input is calculated to replace it in this case. If `net_dict['poisson_input']=False`, the script should replace by a constant DC input. See also the PyNN version of the microcircuit.

jhnnsnk commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                             'withtime': True,
+                             'to_memory': False,
+                             'to_file': True,
+                             'label': self.data + 'volt',
```
for consistency, use 'voltmeter' in the label

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+        if 'spike_detector' in self.net_dict['rec_dev']:
+            print 'spike_detectors created'
+        if 'voltmeter' in self.net_dict['rec_dev']:
+            print 'voltmeter created'
```
--> "voltmeters created."

jhnnsnk commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+        print 'Data will be written to %s' % self.data
+
+    # hand parameters to NEST-kernel
+    def setup(self):
```
please replace the comment by a short docstring here and in the following functions

mschmidt87 replied:
Since this function only sets up NEST, please consider renaming it to `setup_nest`.

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                spikedetector = nest.Create('spike_detector', params=recdict)
+                self.spikedetector.append(spikedetector)
+            if 'voltmeter' in self.net_dict['rec_dev']:
+                rcdictmem = {'interval': self.sim_dict['rec_V_int'],
```
To be consistent, please rename to `recdictmem`

jhnnsnk commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+            rate_ext = self.net_dict['bg_rate'] * self.K_ext
+            self.poisson = []
+            for i, target_pop in enumerate(self.pops):
+                    poisson = nest.Create('poisson_generator')
```
indentation should be 4 spaces

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+    def create_connections(self):
+        mean_delays = self.net_dict['mean_delay_matrix']
+        std_delays = self.net_dict['std_delay_matrix']
+        norm_clp = 'normal_clipped'
```
This variable definition seems obsolete. You can just directly insert 'normal_clipped' in all the dictionaries below.

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                                        'delay': {'distribution': norm_clp,
+                                                  'mu': 1.5,
+                                                  'sigma': 0.75,
+                                                  'low': self.sim_resolution}}
```
`conn_dict_poisson` and `syn_dic_poisson` are not used in this function and can thus be removed.

jhnnsnk commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                                                  'low': self.sim_resolution}}
+                    conn_dict_rec = {'rule': 'fixed_total_number',
+                                     'N': synapse_nr}
+                    syn_dict_rec_ex = {'model': 'static_synapse',
```
you could use only one definition of a `syn_dict` and update it dependent of the sign of the weight:

     syn_dict = {'model': 'static_synapse',
                      'weight': {'distribution': norm_clp,
                                    'mu': weight,
                                    'sigma': w_sd}
                       'delay': {'distribution': norm_clp,
                                   'mu': mean_delays[i][j],
                                   'sigma': std_delays[i][j],
                                   'low': self.sim_resolution}}
    if weight < 0:
        syn_dict['weight']['high'] = 0.0
    else:
        syn_dict['weight']['low'] = 0.0

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                                          'low': 0.0},
+                               'delay': {'distribution': 'normal_clipped',
+                                         'mu': self.stim_dict['delay_th'][i],
+                                         'sigma': stim_dict['delay_th_sd'][i],
```
Please change to `self.stim_dict`

jhnnsnk commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                    nest.Connect(source_pop, target_pop,
+                                 conn_spec=conn_dict_rec,
+                                 syn_spec=syn_dict_rec)
+        print 'Recurrent connections established'
```
I prefer to put such print statements in the beginning of the functions for keeping track of where time is spent, e.g.,: `Establishing recurrent connections`

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+
+    # recording devices are created in this function
+    def create_devices(self):
+        self.spikedetector = []
```
to be consistent with the NEST name, please rename to `self.spike_detector`

jhnnsnk commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+    # poisson generators connected to the different populations
+    def connect_poisson(self):
+        for i, target_pop in enumerate(self.pops):
+                conn_dict_poisson = {'rule': 'all_to_all'}
```
4 spaces, see also the following for-loops

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+                nest.Connect(target_pop, self.spikedetector[i])
+        print '%s of 2 Devices connected' % (len(self.net_dict['rec_dev']))
+
+    def connect(self):
```
Since this function sets up the entire network (incl. creating the populations), please consider renaming it to `setup`.

mschmidt87 commented on pynest/examples/pynest-microcircuit/network.py:
```diff
@@ -0,0 +1,347 @@
+        ts = dSD['times']
+        plt.figure(1)
+        plt.plot(ts, evs, '.')
+        plt.show()
```
In the current state, this function does not work:
- self.spikedetector is a list of tuples of GIDs. `nest.GetStatus` needs a list or a tuple of GIDs so you should adapt self.spikedetector accordingly.
- currently, the spike detector to not record to memory so that the status dictionaries of the spike detectors do not contain the spikes
- Please rewrite this plot function so that it returns a raster plot with properly stacked populations with different colors for E and I populations (like in the manuscript) and a bar plot with the population rates. You can take the PyNN microcircuit as a template.
- Finally, `example.py` should execute the plot function by default.

mschmidt87 commented:
- Currently, this example is not Python3 compatible. We think that it would be good to remedy this. In principle, this should not be too much work . You can use the Python script `2to3` to convert the scripts, which should fix most issues (like the `print` statements), like so `2to3 network.py`.

- In general, it would be good to add some explanation of the different options for the external input to the README file.

- You could add an option to print the time progress (Kernel option `print_time`) on the standard output. Please add a comment then that this should only be used if the network is run on a laptop.

- Currently, membrane potentials are recorded per default. This will drastically slow the simulation at full scale, so it would be better to switch this off per default.

- We haven't tested the simulation results so far. We will do this once you added a plotting/analysis function, that gives a plot of the results incl. the population rates.

This is great work, and most of the issues should not be too much work to fix.

hendrikrth commented:
I misunderstood the comment about the missing Poissonian input, I changed my first version and am now calculating the DC input if no Poissonian input is provided. I also changed the directory name to 'Potjans_2014'.

jhnnsnk commented:
Hi,
thanks for addressing our comments. Here are some remarks on the issues you mention:

1) I have included the exception of the relative standard deviation of the mean weights of the connection L23E to L4E, because this was also done in the PyNN version.
- There was one mistake in your comment "# The weight of the connection L23E to L4E is doubled." The connection that is doubled is from L4E to L23 and not the other way round. In the code, it was correct.
- What we criticized initially was that in your version this value is hard-coded in the main code and not in the parameter files. Since the resulting activity does not differ a lot for 5 or 10 %, we just agreed with Sacha that it is probably easier to use 10 % also for this connection as done in the SLI version.

2) I don't create the output dir automatically, because if the simulation is run with MPI, this can cause problems.
- You can let only the first MPI process create an output directory, using `if nest.Rank() == 0` .

3) I have implemented your feedback about initiating the random membrane potentials with the sli command. But I am not sure if this is save to do, since I am using the command nest.Rank() to get the MPI Rank for the random number state. In the NEST documentation it says, that you should only use this command for debugging purposes. Is this still the case, or do you know another way of getting the Rank of the MPI process.
- Better not use nest.Rank() here. We suggest to call `vp = nest.GetStatus(local_nodes)[0]['vp']` in order to get the virtual process. vp is the same for all your local nodes - they are local to the virtual process and not only the MPI process. Then, you can use `nest.SetStatus()` with `self.pyrngs[vp]` similar to what you already have.

4) You wrote, that all members should be initialized in the init function. Do you mean all self variables, that I define in the outer functions?
- We had a look again and just agreed that you can leave it as it is.

5) You wrote that I should include in the function 'adj_w_ext_to_K' in helpers.py an option if no Poisson input is provided. This was already implemented in the else. But I made it clearer by changing the else to 'elif net_dict['poisson_input'] is not True and stim_dict['dc_input']:'.
- We did not see this in your latest version. We will add comments to this function when we have  a closer look at the code again.

6) I have added the default option, that the mean and std of the firing rates are calculated, a spike raster plot and a boxplot are created. But so far this is only possible, if the simulation is run on with one mpi-process and one thread per mpi-process. Is this sufficient? 
- As above, you can use `if nest.Rank() == 0`.

We will check out the code again after you addressed the points above, ok?
A final remark: Please reply directly to our line comments - that makes it much easier to follow.

hendrikrth commented:
Hi, 
I have inherited your comments about the mistake in the comment (...weight of L23E to L4E...) and changed it. I removed the exception for the standard deviation, so that all standard deviations are 10 % now (-->1).
I also inherited your feedback about the initialization of the membrane potentials (-->3).
Adressing points 2 and 6:
I still believe that creating the dir and calculating the firing rates, creating the raster and box plot only if nest.Rank() == 0 is not a save way to do. My concerns come from the fact, that the MPI process creating the dir might be slower than the others. If this happens the other MPI process might try to store files in a dir, which has not been created yet. My second concern is that if only one nest.Rank() == 0 plots the results and the others haven't finished yet, there might be some lacking information. 
Therefore I propose, that I write a bash script in which the user defines if the simulation should be run with MPI process and defines also the number of them. This script then creates the output directory and executes after finishing the simulation either with MPI or without an analyze script. What are your thoughts about this?

mschmidt87 commented:
Thanks for addressing our points.

> I still believe that creating the dir and calculating the firing rates, creating the raster and box plot only if nest.Rank() == 0 is not a save way to do. My concerns come from the fact, that the MPI process creating the dir might be slower than the others. If this happens the other MPI process might try to store files in a dir, which has not been created yet. My second concern is that if only one nest.Rank() == 0 plots the results and the others haven't finished yet, there might be some lacking information. 

The MPI ranks are synchronized in NEST at the beginning and the end of the simulation. So, this should not be a problem.

heplesser commented:
@mschmidt87 @jhnnsnk From your most recent comments, I am not quite sure whether you approve the PR as it now is or whether you still require changes. Could you use the "Review Changes" button in the "Files changed" tab and check off whether you approve or require changes?

mschmidt87 commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,475 @@
+                        )
+
+    def connect_poisson(self):
+        """ Connects the Poisson generator to the microcircuit."""
```
change to "Poisson generators" to avoid the impression that all neurons share one generator.

hendrikrth replied:
Accepted

mschmidt87 commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,475 @@
+        self.connect_devices()
+
+    def simulate(self):
+        """ This function starts the simulation."""
```
It does not only start the simulation, thus it should read e.g.
""" This function simulates the network."""

hendrikrth replied:
Accepted and changed

mschmidt87 commented on pynest/examples/Potjans_2014/example.py:
```diff
@@ -0,0 +1,54 @@
+from stimulus_params import stim_dict
+
+
+# initialize the network and pass parameters to it
```
To be consistent through all script, please capitalize the first word of each comment here.

mschmidt87 commented on pynest/examples/Potjans_2014/example.py:
```diff
@@ -0,0 +1,54 @@
+net.simulate()
+time_sim = time.time()-t2
+print("Time to simulate: %.2f seconds" % time_sim)
+# plot results of the simulation
```
Some more explanation would be nice here, e.g.
" Plot a raster plot of the simulated spikes and the average spike rates of all populations. For visual purposes, the last 200 ms are plotted here by default. To  compute the spike rates, the first 500 ms of the simulation are discarded to neglect initial transients."

mschmidt87 commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -0,0 +1,462 @@
+    for h in list(range(len(files))):
+        n_fil = data_all[h][:, 0]
+        n_fil = n_fil.astype(int)
+        t_fil = data_all[h][:, 1]
```
t_fil is introduced but never used.

hendrikrth replied:
I removed this variable

hendrikrth commented on pynest/examples/Potjans_2014/readme.md:
```diff
@@ -0,0 +1,32 @@
+How to use the example:
+
+To run the microcircuit on a local machine, adjust the variables `N_scaling` and `K_scaling` in `network_params.py` to 0.1 and create an output directory called `data`. `N_scaling` adjusts the number of neurons and `K_scaling` the number of connections to be simulated. The full network can be run by adjusting these values to 1. If this is done, the option to print the time progress should be set  ...
+The code can be parallelized using OpenMP and MPI, if NEST has been built with these applications [Parralel computing with NEST](http://www.nest-simulator.org/parallel_computing/). The number of threads (per MPI process) can be chosen by adjusting `local_num_threads` in `sim_params.py`. The number of MPI process can be set by choosing a reasonable value for `num_mpi_prc` and then running the scri ...
```
I incorporated your comments.

hendrikrth commented on pynest/examples/Potjans_2014/stimulus_params.py:
```diff
@@ -0,0 +1,65 @@
+    'dc_input': False,
+    # Number of thalamic neurons
+    'n_thal': 902,
+    # Amplitude of the thalamic postsynaptic potential (in mV)
```
Accepted

hendrikrth commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -0,0 +1,246 @@
+
+    The function creates a matrix of the mean evoked postsynaptic
+    potentials between the recurrent connections of the microcircuit.
+    The weight of the connection between L4E and L23E is doubled.
```
Accepted

hendrikrth commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -0,0 +1,246 @@
+    'K_scaling': 0.1,
+    # Factor to scale the number of neurons.
+    'N_scaling': 0.1,
+    # Amplitude of excitatory postsynaptic potential (in mV).
```
Accepted

hendrikrth commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,475 @@
+import nest
+import numpy as np
+import os
+import sys
```
Accepted and removed

hendrikrth commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,475 @@
+
+        This function resets the NEST-kernel and passes parameters to it.
+        The number of seeds for the NEST-kernel is computed, based on the
+        total number of mpi-processes and threads of each.
```
Done

hendrikrth commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,475 @@
+            )
+
+    def create_devices(self):
+        """ This function creates the recording devices."""
```
Accepted

hendrikrth commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,475 @@
+
+        If thalamic input should be simulated, a thalamic neuronal population
+        is created.
+
```
Accepted

hendrikrth commented on pynest/examples/Potjans_2014/readme.md:
```diff
@@ -0,0 +1,32 @@
+	Contains the parameters for the stimuli.
+	* `example.py`  
+   Use this script to try out the microcircuit.
+   
```
I added a description

mschmidt87 commented on pynest/examples/Potjans_2014/stimulus_params.py:
```diff
@@ -0,0 +1,74 @@
+    }
+
+
+stim_dict.update(updated_dict)
```
I don't understand the newly added construction with `updated_dict` here. In contrast to `net_dict` in `network_params.py`, this seems unnecessary here, because you don't use any parameters from `stim_dict` in `updated_dict`.

hendrikrth replied:
Yes, you are right that made no sense. I removed the update routine.

mschmidt87 commented:
:+1:  from me.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,477 @@
+            'Seeds for random number generators of virtual processes: %r'
+            % rng_seeds
+            )
+        print('Global random number generator seed:%i' % grng_seed)
```
Add a space in front of `%i` for being consistent.

hendrikrth replied:
Done

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,477 @@
+
+    def connect_devices(self):
+        """ Connects the recording devices to the microcircuit."""
+        print('%s of 2 Devices connected' % (len(self.net_dict['rec_dev'])))
```
It prints "1 of 2 Devices connected" if only a spike detector is used which sounds as if something was missing. I suggest to be specific and print "Spike Detector connected" or/and "Voltmeter connected".

hendrikrth replied:
fixed

jhnnsnk commented on pynest/examples/Potjans_2014/example.py:
```diff
@@ -0,0 +1,57 @@
+
+
+# Initialize the network and pass parameters to it.
+net = network.Network(sim_dict, net_dict, stim_dict)
```
It would be nice to time also the network initialization.

hendrikrth replied:
Done

jhnnsnk commented on pynest/examples/Potjans_2014/example.py:
```diff
@@ -0,0 +1,57 @@
+t1 = time.time()
+net.setup()
+time_con = time.time()-t1
+print("Time to create the connections: %.2f seconds" % time_con)
```
Since the units `ms` and `Hz` are printed as well, why not just writing `s` instead of `seconds`?

hendrikrth replied:
Done

jhnnsnk commented on pynest/examples/Potjans_2014/readme.md:
```diff
@@ -0,0 +1,35 @@
+   
+How to use the example:
+
+To run the microcircuit on a local machine, adjust the variables `N_scaling` and `K_scaling` in `network_params.py` to 0.1 and create an output directory called `data`. `N_scaling` adjusts the number of neurons and `K_scaling` the number of connections to be simulated. The full network can be run by adjusting these values to 1. If this is done, the option to print the time progress should be set  ...
```
Remove 

> and create an output directory called `data`. 

since this is now done automatically.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,477 @@
+        """ This function simulates the microcircuit."""
+        nest.Simulate(self.sim_dict['t_sim'])
+
+    def evaluate(self, raster_plot_time_idx, fire_rate_time_idx):
```
You could save the generated plots directly in the output directory together with the raw data.

jhnnsnk commented on pynest/examples/Potjans_2014/stimulus_params.py:
```diff
@@ -0,0 +1,67 @@
+    # Standard deviation of the postsynaptic potential (in relative units).
+    'PSP_sd': 0.1,
+    # Start of the thalamic input (in ms).
+    'th_start': 700.0,
```
It would be nice to set the default stimulus intervals somewhere into the default plotting interval so that their effect is directly visible by eye.

hendrikrth replied:
I kept the thalamus stimulus interval, since this is the stimulus interval of the sli version. I therefore changed the interval of the spike raster plot.

jhnnsnk replied:
Ok. Please update the comment on the plotting interval accordingly.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,477 @@
+            else:
+                os.mkdir(self.sim_dict['data_path'])
+                print('data directory created')
+        print('Data will be written to %s' % self.data_path)
```
If used with MPI such statements are currently printed once per process. Please print them only on Rank 0 unless you want to print something specific to a process.

hendrikrth replied:
Done

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,477 @@
+            '0 << /model /%s >> GetLocalNodes' % self.net_dict['neuron_model']
+            )
+        local_nodes = nest.sli_pop()
+        vp = nest.GetStatus(local_nodes)[0]['vp']
```
If run with both MPI and threads, taking the vp of the first local node is not correct. For example, with 2 MPI processes and 2 threads, one gets the MPI ranks 0 and 1, but the vps 0,1,2 and 3. For accessing the right pyrngs, I think we cannot avoid to extract all vps correponding to the local nodes.

jhnnsnk replied:
This might be a possible solution to the problem:

```python
for thread in np.arange(nest.GetKernelStatus('local_num_threads')):
    local_nodes = nest.GetNodes([0], {'model': self.net_dict['neuron_model'], 'thread': thread}, local_only=True)[0]
    vp = nest.GetStatus(local_nodes)[0]['vp'] # same thread on MPI proc -> same vp
    nest.SetStatus(
        local_nodes, 'V_m', self.pyrngs[vp].normal(
        self.net_dict['neuron_params']['V0_mean'],
        self.net_dict['neuron_params']['V0_sd'],
            len(local_nodes)))
```

hendrikrth replied:
I fixed this, using your solution.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,477 @@
+            self.DC_amp_e = compute_DC(self.net_dict, self.w_ext)
+
+        print(
+            'The Number of neurons is scaled by a factor of: %.2f'
```
Number -> number

hendrikrth replied:
Done

jhnnsnk commented:
Hi, during testing with both MPI and threads, I noticed that the initialization of membrane voltages did not work correctly. I made a suggestion inline how this could be solved. Apart from this, I made some other minor comments for improvement.

mschmidt87 commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,506 @@
+        nest.SetKernelStatus(kernel_dict)
+
+    def create_populations(self):
+        """ This function creates the neuronal populations.
```
see above, please rephrase to "Creates..."
please repeat this for all other functions.

mschmidt87 commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,506 @@
+                     }, local_only=True
+                )[0]
+            vp = nest.GetStatus(local_nodes)[0]['vp']
+            # same thread on MPI proc -> same vp
```
I don't quite understand this comment, can you elaborate what you mean with this?

hendrikrth replied:
I adopted this comment and the code to initialize the membrane potential from jhnnsnk. This comment means - if I understand it correctly (and the code above) - that the nest.GetStatus command is called for all local nodes in one thread. Then the vp of the first node is extracted to initialize the membrane potential. This is safe, because only local nodes on the same thread are retrieved. Then this is done for the next thread. 
So the comment should maybe elaborate this a little bit more: 
--> vp the same for all local nodes on the same thread

mschmidt87 commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -0,0 +1,247 @@
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+'''
+pynest microcicuit parameters
```
--> microcircuit

mschmidt87 commented on pynest/examples/Potjans_2014/readme.md:
```diff
@@ -0,0 +1,35 @@
+	* `helpers.py`  
+	Helper functions for the simulation and evaluation of the microcircuit.	
+	* `network.py`  
+	File which gathers all parameters and connects the different nodes with each other.
```
Rephrase to: "Gathers..." in acoordance with the descriptions of the other files.

mschmidt87 commented on pynest/examples/Potjans_2014/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+'''
+microcicuit simulation parameters
```
microcircuit

mschmidt87 commented on pynest/examples/Potjans_2014/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+microcicuit simulation parameters
+---------------------------------
+
+Simulation Parameters for the microcircuit.
```
Simulation parameters

mschmidt87 commented on pynest/examples/Potjans_2014/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+    'local_num_threads': 1,
+    # Recording interval of the membrane potential (in ms).
+    'rec_V_int': 1.0,
+    # If True data will be overwritten,
```
If True, data...

mschmidt87 commented on pynest/examples/Potjans_2014/sim_params.py:
```diff
@@ -0,0 +1,51 @@
+    # Recording interval of the membrane potential (in ms).
+    'rec_V_int': 1.0,
+    # If True data will be overwritten,
+    # if False a NESTError is raised if the files already exist.
```
If False, ...

mschmidt87 commented on pynest/examples/Potjans_2014/stimulus_params.py:
```diff
@@ -0,0 +1,67 @@
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+'''
+microcicuit stimulus parameters
```
microcircuit

hendrikrth commented on pynest/examples/Potjans_2014/example.py:
```diff
@@ -0,0 +1,63 @@
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+'''
+pynest microcicuit example
```
fixed this and all other typos

hendrikrth commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -0,0 +1,466 @@
+
+
+def get_weight(PSP_val, net_dict):
+    """ Function computes weight to elicit a change in the membrane potential.
```
Changes this and adapted this style in all other functions

hendrikrth commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -0,0 +1,466 @@
+        )
+    # If the network is scaled the indegrees are calculated in the same
+    # fashion as in the original version of the circuit, which is
+    # written in sli
```
Done

hendrikrth commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -0,0 +1,466 @@
+def adj_w_ext_to_K(K_full, K_scaling, w, w_from_PSP, DC, net_dict, stim_dict):
+    """ Adjustment of weights to scaling is dine in this function.
+
+    With this funtion the recurrent and external weights are adjusted
```
Done

hendrikrth commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -0,0 +1,466 @@
+
+
+def load_spike_times(path, name, begin, end):
+    """ This function loads spike times of each spike detector.
```
Done

hendrikrth commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -0,0 +1,506 @@
+        The number of seeds for the NEST-kernel is computed, based on the
+        total number of MPI processes and threads of each.
+
+        Parameters
```
Done

jhnnsnk commented:
:+1:

heplesser commented:
@hendrikrth Thank you for porting this important model to PyNEST! @jhnnsnk @mschmidt87 Thank you for your thorough review!
